### PR TITLE
Pass search item to icon component, pass icon as component to search list item

### PR DESCRIPTION
**Pull request by IlyaSavich**
Make it possible to customize icon component by search item details

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by camilaibs — COMMENTED** [review]
Thanks for contributing ⭐ 
Let me understand the necessity first.
What you're trying to do is customize the icon component based on the result?

If that's the case, my suggestion is to keep the `icon` property as it is today, without renaming it, and modify its type to accept a ReactNode or a render function, without introducing breaking changes.

 It's still going to work like it does now:
```tsx
<CatalogSearchResultListItem icon={<CatalogIcon />} />
```
People will also be able to pass a render function to the icon property if they want to customize it:
```tsx
<CatalogSearchResultListItem
  icon={result => {
    if (result.type === "api") {
      return (
        <ListItemIcon>
          <ApiIcon />
        </ListItemIcon>
      );
     }
    return (
      <ListItemIcon>
        <CatalogIcon />
      </ListItemIcon>
    );
  }
/>
```
What do you think?

**Review by camilaibs — APPROVED** [review]
LGTM 🎉 , thanks again for contributing!

File: .changeset/slow-insects-cheat.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+'@backstage/plugin-techdocs': minor
+'@backstage/plugin-catalog': minor
+'@backstage/plugin-explore': minor
+'@backstage/create-app': patch
```
Comment (freben) [inline .changeset/slow-insects-cheat.md:5]: Let's remove this line.

**Review by freben — COMMENTED** [review]
Excellent! Let's just remove the extra line in the changeset and then :shipit:

File: .changeset/slow-insects-cheat.md (hunk 1 above)
Comment (IlyaSavich) [reply .changeset/slow-insects-cheat.md]: done

File: .changeset/slow-insects-cheat.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+'@backstage/plugin-techdocs': minor
+'@backstage/plugin-catalog': minor
+'@backstage/plugin-explore': minor
```
Comment (benjdlambert) [inline .changeset/slow-insects-cheat.md:4]: Think this should be `patch` for the `explore` plugin as it's <1.0.0. But the rest are fine

Comment (IlyaSavich) [reply .changeset/slow-insects-cheat.md]: changed to patch for all of them as there the same changes

**Comment by benjdlambert** [comment]
@IlyaSavich the other ones should still be minor, as they're v1.0. I updated the changeset to fix it all, so once it's green we should be able to merge.

**Pull request merged**